lib/krane/deploy_task.rb

```diff
@@ -1,12 +1,394 @@
+    def predeploy_sequence
+      before_crs = %w(
+        ResourceQuota
+        NetworkPolicy
+      )
+      after_crs = %w(
+        ConfigMap
+        PersistentVolumeClaim
+        ServiceAccount
+        Role
+        RoleBinding
+        Secret
+        Pod
+      )
+
+      before_crs + cluster_resource_discoverer.crds.select(&:predeployed?).map(&:kind) + after_crs
+    end
+
+    def prune_whitelist
+      wl = %w(
+        core/v1/ConfigMap
+        core/v1/Pod
+        core/v1/Service
+        core/v1/ResourceQuota
+        core/v1/Secret
+        core/v1/ServiceAccount
+        core/v1/PodTemplate
+        core/v1/PersistentVolumeClaim
+        batch/v1/Job
+        apps/v1/ReplicaSet
+        apps/v1/DaemonSet
+        apps/v1/Deployment
+        extensions/v1beta1/Ingress
+        networking.k8s.io/v1/NetworkPolicy
+        apps/v1/StatefulSet
+        autoscaling/v1/HorizontalPodAutoscaler
+        policy/v1beta1/PodDisruptionBudget
+        batch/v1beta1/CronJob
+        rbac.authorization.k8s.io/v1/Role
+        rbac.authorization.k8s.io/v1/RoleBinding
+      )
+      wl + cluster_resource_discoverer.crds.select(&:prunable?).map(&:group_version_kind)
+    end
+
+    def server_version
+      kubectl.server_version
+    end
+
+    # Initializes the deploy task
+    #
+    # @param namespace [String] Kubernetes namespace
+    # @param context [String] Kubernetes context
+    # @param current_sha [String] The SHA of the commit
+    # @param logger [Object] Logger object (defaults to an instance of Krane::FormattedLogger)
+    # @param kubectl_instance [Kubectl] Kubectl instance
+    # @param bindings [Hash] Bindings parsed by Krane::BindingsParser
+    # @param max_watch_seconds [Integer] Timeout in seconds
+    # @param selector [Hash] Selector(s) parsed by Krane::LabelSelector
+    # @param template_paths [Array<String>] An array of template paths
+    # @param protected_namespaces [Array<String>] Array of protected Kubernetes namespaces (defaults
+    #   to Krane::DeployTask::PROTECTED_NAMESPACES)
+    # @param render_erb [Boolean] Enable ERB rendering
+    def initialize(namespace:, context:, current_sha:, logger: nil, kubectl_instance: nil, bindings: {},
+      max_watch_seconds: nil, selector: nil, template_paths: [], protected_namespaces: nil,
+      render_erb: true)
+      template_paths = (template_paths.map { |path| File.expand_path(path) }).compact
+
+      @logger = logger || Krane::FormattedLogger.build(namespace, context)
+      @template_sets = TemplateSets.from_dirs_and_files(paths: template_paths, logger: @logger)
+      @task_config = Krane::TaskConfig.new(context, namespace, @logger)
+      @bindings = bindings
+      @namespace = namespace
+      @namespace_tags = []
+      @context = context
+      @current_sha = current_sha
+      @kubectl = kubectl_instance
+      @max_watch_seconds = max_watch_seconds
+      @selector = selector
+      @protected_namespaces = protected_namespaces || PROTECTED_NAMESPACES
```

Comment:
Should `PROTECTED_NAMESPACES` be the default in the function signature instead of `nil`

Reply:
Doing that doesn't work for some reason; a bunch of tests break. I'll check the other things and update this last if I can.

Reply:
Turns out this is because we explicitly pass `nil` from `deploy_fixtures`